### ISSUE TITLE
How to (or should we) use context managers in path operators?  Specifically, with StreamingResponses?

### ISSUE THREAD

**Issue by falkben** [tiangolo/fastapi#1936]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

Context managers could be used to close a database or close a file.  I'm having a bit of trouble using them with fastapi path operators, as the finally block in the context manager doesn't seem to be executed in the order I would expect unless you use the context manager as a fastapi dependency.

### Example

```Python
import time
from contextlib import contextmanager

import uvicorn
from fastapi import FastAPI
from fastapi.params import Depends
from fastapi.responses import StreamingResponse

app = FastAPI()


@contextmanager
def simple_context_manager():
    """ this could be a wrapper around a database """

    def _gen():
        iterations = 1000
        for i in range(iterations):
            yield str(i)
            time.sleep(0.001)
            if i == iterations - 1:
                print("reached the end")

    try:
        yield _gen
    finally:
        # we could be closing the database conn here
        print("we are cleaning up")


def mock_db_query():
    """ Dependency wrapping a context manager """
    with simple_context_manager() as db:
        yield db


@app.get("/stream_context_mngr")
async def stream_context_mngr():
    with simple_context_manager() as gen:
        # ! this causes the finally block in the context manager to occur immediately *before* streaming the response
        # returning StreamingResponse causes the context manager to close for some reason
        return StreamingResponse(gen())


@app.get("/stream_dep")
async def stream_dep(db=Depends(mock_db_query)):
    """ could be sync or async path operator here 
    This works as expected.  Finally block in the "wrapped" context manager occurs after the response is finished """
    return StreamingResponse(db())


@app.get("/direct_context_mngr")
async def direct_context_mngr():
    with simple_context_manager() as gen:
        # This works
        return gen()


if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)
```

### Description

I'm using `print` statement to the console to observe the order in which things are executed.  In this case I would expect that "reached the end" would occur before the "we are cleaning up" print in all cases.

However, when "/stream_context_mngr" are called we perform the cleanup stage in the finally block *before* consuming the generator.

It seems like the context manager might be closed as soon as we hit the `return StreamingResponse`?

In the [docs](https://fastapi.tiangolo.com/advanced/custom-response/#using-streamingresponse-with-file-like-objects) a context manager is specifically *not* used to open a file w/ StreamingResponse.  If you do, it fails.  I understand that the file should be closed once the file handle goes out of scope, but maybe there are some cases where a context manager can be helpful.

My questions are as follows:

- Should we not use context managers with FastAPI? 
- Why is it that we can't use context managers? 
- Should this be added to the docs?

### Environment

* OS: Linux & MacOS:
* FastAPI Version: 0.60.1
* Python version: 3.8.3

EDIT: I had incorrectly stated that returning the response directly (without streaming response) also caused the finally block to execute first. However, that does not seem to be the case now. It seems to me now that context managers are generally just not compatible with `StreamingResponse` but that you can use them elsewhere in fastapi.

**Comment by falkben** [comment]
Yes, this toy example will work, but my point is that the finally block executes before you start streaming the response.  Imagine if you had a file handle or a database connection.  It would close the file or connection before attempting to iterate through the generator. Dependencies allow you to perform work after the return, but they do have some drawbacks regarding exception handling.

My question is whether I have everything correct about context managers and where they fit in the response cycle.  Secondarily, perhaps we could write some documentation to explain this "gotcha" about using context managers. (I'd be happy to take a first stab at writing it)

**Comment by ihakh** [comment]
your syntax was wrong
you must pass the generator without the parentheses look at the documentation:
https://www.starlette.io/responses/#streamingresponse
this is official example:
```python
from starlette.responses import StreamingResponse
import asyncio


async def slow_numbers(minimum, maximum):
    yield('<html><body><ul>')
    for number in range(minimum, maximum + 1):
        yield '<li>%d</li>' % number
        await asyncio.sleep(0.5)
    yield('</ul></body></html>')


async def app(scope, receive, send):
    assert scope['type'] == 'http'
    generator = slow_numbers(1, 10)
    response = StreamingResponse(generator, media_type='text/html')
    await response(scope, receive, send)
```
I think you did not read the document thoroughly
what I have done is to remove the parentheses from generator when passing it to the `StreamingResponse`
do it in your actual code see if this solves the problem

**Comment by ihakh** [comment]
the contextmanager does the finish stuff after the call to the managed resource is finished
you are calling the managed resource before passing it to the StreamingResponse
so when the call is finished, finishing stuff is called
actually the managed resource (e.g. the generator must be called inside the StreamingResponse so when the StreamingResponse job is done the contextmanager does the finishing stuff

**Comment by falkben** [comment]
Thanks for the comment @ihakh . Hopefully the following will clear things up a little

In the example I posted, `gen` without parentheses is just a function handle. If you don't call it, StreamingResponse tries to call `next` on `gen` and it raises an exception. To use StreamingResponse you need to pass it a generator (or async gen, same thing though). To get a generator in my example, you need to call the function, that's why the parentheses are there. Perhaps clearer for me to have named it `gen_fcn` or something. In your example, you do the exact same, with `generator = slow_numbers(1, 10)`, it's just that you use arguments there.

Also, it doesn't matter if you return a generator or a function handle to create the generator.  Swapping out `yield gen()` instead of `yield gen` in `simple_context_manager` (and then just pass `gen` into StreamingResponse), doesn't change anything -- the problem persists, the `finally` code still executes before you start iterating through the generator.

Here's another example showing this behavior re: StreamingResponse & context managers:

```py
@app.get("/file_context_mngr_stream")
async def file_context_mngr_stream():
    with open(f"{__file__}") as f:
        return StreamingResponse(f)
```

I know that in the [docs](https://fastapi.tiangolo.com/advanced/custom-response/#using-streamingresponse-with-file-like-objects) it shows a different way to return files (either using FileResponse or to not use a context manager). My question is just that if this is known behavior regarding context managers and streaming responses, should we mention it in the docs, and can someone perhaps lead me towards *why* this is happening in the first place.  It is a minor "gotcha" re: fastapi that perhaps we can avoid some confusion for newcomers like myself.

**Comment by falkben** [comment]
This is now answered in the docs. (maybe it was before and I missed it?)

https://fastapi.tiangolo.com/tutorial/dependencies/dependencies-with-yield/#using-context-managers-in-dependencies-with-yield

if you need or want to use a context manager with a `StreamingResponse`, just put it *inside* a `Depends` 

Very nice.

**Comment by tiangolo** [comment]
Thanks for the discussion here! And thanks for coming back to close the issue @falkben ☕ 

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
